src/os/bluestore/BitmapFreelistManager.cc

```diff
@@ -0,0 +1,399 @@
+	if (!enumerate_p->valid()) {
+	  break;
+	}
+	string k = enumerate_p->key();
+	const char *p = k.c_str() + 1;
+	_key_decode_u64(p, &enumerate_offset);
+	enumerate_bl = enumerate_p->value();
+      }
+    }
+  }
+
+  end = blocks * bytes_per_block;
+  if (enumerate_offset < end) {
+    *length = end - enumerate_offset;
+    dout(10) << __func__ << " " << *offset << "~" << *length << dendl;
+    enumerate_offset = end;
+    return true;
+  }
+
+  dout(10) << __func__ << " end" << dendl;
+  return false;
+}
+
+void BitmapFreelistManager::dump()
+{
+  enumerate_reset();
+  uint64_t offset, length;
+  while (enumerate_next(&offset, &length)) {
+    dout(20) << __func__ << " " << offset << "~" << length << dendl;
+  }
+}
+
+int BitmapFreelistManager::allocate(
+  uint64_t offset, uint64_t length,
+  KeyValueDB::Transaction txn)
+{
+  dout(10) << __func__ << " " << offset << "~" << length << dendl;
+  return _xor(offset, length, txn);
+}
+
+int BitmapFreelistManager::release(
+  uint64_t offset, uint64_t length,
+  KeyValueDB::Transaction txn)
+{
+  dout(10) << __func__ << " " << offset << "~" << length << dendl;
+  return _xor(offset, length, txn);
+}
+
+int BitmapFreelistManager::_xor(
+  uint64_t offset, uint64_t length,
+  KeyValueDB::Transaction txn)
+{
+  // must be block aligned
+  assert((offset & block_mask) == offset);
+  assert((length & block_mask) == length);
+
+  uint64_t first_key = offset & key_mask;
+  uint64_t last_key = (offset + length - 1) & key_mask;
+  dout(20) << __func__ << " first_key " << first_key << " last_key " << last_key
+	   << dendl;
+
+  if (first_key == last_key) {
+    bufferptr p(blocks_per_key >> 3);
+    p.zero();
+    unsigned s = (offset & ~key_mask) / bytes_per_block;
+    unsigned e = ((offset + length - 1) & ~key_mask) / bytes_per_block;
+    for (unsigned i = s; i <= e; ++i) {
+      p[i >> 3] ^= 1ull << (7 - (i & 7));
+    }
+    string k;
+    make_offset_key(first_key, &k);
+    bufferlist bl;
+    bl.append(p);
+    txn->merge(prefix, k, bl);
+  } else {
+    // first key
+    {
+      bufferptr p(blocks_per_key >> 3);
+      p.zero();
+      unsigned s = (offset & ~key_mask) / bytes_per_block;
+      unsigned e = blocks_per_key;
+      for (unsigned i = s; i < e; ++i) {
+	p[i >> 3] ^= 1ull << (7 - (i & 7));
+      }
+      string k;
+      make_offset_key(first_key, &k);
+      bufferlist bl;
+      bl.append(p);
+      dout(30) << __func__ << " " << first_key << ": ";
+      bl.hexdump(*_dout);
+      *_dout << dendl;
+      txn->merge(prefix, k, bl);
+      first_key += bytes_per_block * blocks_per_key;
+    }
+    // middle keys
+    if (first_key < last_key) {
+      bufferptr p(blocks_per_key >> 3);
+      memset(p.c_str(), 0xff, p.length());
+      while (first_key < last_key) {
+	string k;
+	make_offset_key(first_key, &k);
+	bufferlist bl;
+	bl.append(p);
+	dout(30) << __func__ << " " << first_key << ": ";
+	bl.hexdump(*_dout);
+	*_dout << dendl;
+	txn->merge(prefix, k, bl);
```

Comment:
I guess that you probably intend to use the global initialized `all_set_bl` here(to replace bl), for the sake of performance:)